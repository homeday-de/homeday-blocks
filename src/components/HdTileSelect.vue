<template>
  <div class="field tile-select">
    <div class="tile-select__items">
      <hd-tile-select-item
        v-for="item in itemsMapped"
        :key="item.value"
        :value="item.value"
        :text="item.text"
        :model="value"
        @input="$emit('input', $event)"
      />
      <hd-tile-select-editable-item
        v-if="acceptNewValue"
        v-bind="$attrs"
        v-model="customInputModel"
        :text="customInputText"
        :model="value"
      />
    </div>
  </div>
</template>

<script>
/**
 * @typedef ItemEntryObject
 * @type {Object}
 * @property {string}                   text  - Text to be used in the tile
 * @property {(string|number|boolean)}  value - Value associated with the tile
 */

import _get from 'lodash/get';
import HdTileSelectItem from 'homeday-blocks/src/components/HdTileSelectItem.vue';
import HdTileSelectEditableItem from 'homeday-blocks/src/components/HdTileSelectEditableItem.vue';

export default {
  name: 'HdTileSelect',
  inheritAttrs: false,
  components: {
    HdTileSelectItem,
    HdTileSelectEditableItem,
  },
  data() {
    return {
      customInputModel: '',
    };
  },
  props: {
    /** @type {(string|number|boolean|ItemEntryObject)[]} */
    items: {
      type: Array,
      required: true,
      validator(value) {
        return value.length > 1;
      },
    },
    value: {
      type: [String, Number, Boolean],
      default: '',
    },
    formatter: {
      type: Function,
      default: value => (String(value)),
    },
    acceptNewValue: {
      type: Boolean,
      default: false,
    },
  },
  watch: {
    customInputModel(newValue) {
      this.$emit('input', newValue);
    },
  },
  computed: {
    itemsMapped() {
      return this.items.map(item => ({
        value: _get(item, 'value', item),
        text: this.formatter(_get(item, 'text', item)),
      }));
    },
    customInputText() {
      if (this.customInputModel === '') {
        return '+';
      }

      return this.formatter(this.customInputModel);
    },
  },
};
</script>

<style lang="scss" scoped>
@import "homeday-blocks/src/styles/mixins.scss";

$item-min-size: 100px;
$border-width: 1px;

.tile-select {
  @include font('text-xsmall');

  &__items {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax($item-min-size, 1fr));
    grid-gap: $border-width;
    background-color: getShade($quaternary-color, 60);
    border: $border-width solid getShade($quaternary-color, 60);
    border-radius: 4px;
    font-weight: 600;
    overflow: hidden;
  }
}
</style>
