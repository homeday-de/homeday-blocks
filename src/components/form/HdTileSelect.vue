<template>
  <FieldBase
    v-bind="$attrs"
    :name="name"
    :error="error"
    :helper="helper"
    :filled="isFilled"
    minimized-label
    grouped
  >
    <div class="field tile-select">
      <div class="tile-select__items">
        <hd-tile-select-item
          v-for="item in itemsMapped"
          :key="item.value"
          :value="item.value"
          :text="item.text"
          :model="value"
          @input="emitValue"
        />
        <hd-tile-select-editable-item
          v-if="acceptNewValue"
          v-bind="$attrs"
          v-model="customInputModel"
          :text="customInputText"
          :model="value"
        />
      </div>
    </div>
  </FieldBase>
</template>

<script>
/**
 * @typedef ItemEntryObject
 * @type {Object}
 * @property {string}                   text  - Text to be used in the tile
 * @property {(string|number|boolean)}  value - Value associated with the tile
 */

import _get from 'lodash/get';
import _merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import HdTileSelectItem from 'homeday-blocks/src/components/HdTileSelectItem.vue';
import HdTileSelectEditableItem from 'homeday-blocks/src/components/HdTileSelectEditableItem.vue';
import FieldBase from './FieldBase.vue';
import formField from './formFieldMixin';

export default {
  name: 'HdTileSelect',
  inheritAttrs: false,
  mixins: [
    formField,
  ],
  components: {
    HdTileSelectItem,
    HdTileSelectEditableItem,
    FieldBase,
  },
  data() {
    return {
      customInputModel: '',
      error: null,
      helper: null,
    };
  },
  props: {
    /** @type {(string|number|boolean|ItemEntryObject)[]} */
    items: {
      type: Array,
      required: true,
      validator(value) {
        return value.length > 1;
      },
    },
    value: {
      type: [String, Number, Boolean],
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    formatter: {
      type: Function,
      default: value => (String(value)),
    },
    acceptNewValue: {
      type: Boolean,
      default: false,
    },
    required: {
      type: Boolean,
      default: false,
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
  },
  watch: {
    customInputModel(newValue) {
      this.emitValue(newValue);
    },
  },
  computed: {
    t() {
      return _merge(getMessages(this.lang), this.texts);
    },
    itemsMapped() {
      return this.items.map(item => ({
        value: _get(item, 'value', item),
        text: this.formatter(_get(item, 'text', item)),
      }));
    },
    customInputText() {
      if (this.customInputModel === '') {
        return '+';
      }

      return this.formatter(this.customInputModel);
    },
    isFilled() {
      return this.value !== null && this.value !== undefined && this.value !== '';
    },
  },
  methods: {
    async emitValue(value) {
      this.$emit('input', value);
      await this.$nextTick();
      this.validate();
    },
    showError(errorMessage) {
      this.error = errorMessage;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.error = null;
    },
    validate() {
      if (this.required && !this.isFilled) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else {
        this.hideError();
      }
      return !this.error;
    },
  },
};
</script>

<style lang="scss" scoped>
@import "homeday-blocks/src/styles/mixins.scss";

$item-min-size: 100px;
$border-width: 1px;

.tile-select {
  @include font('text-xsmall');

  &__items {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax($item-min-size, 1fr));
    grid-gap: $border-width;
    flex: 1;
    background-color: getShade($quaternary-color, 60);
    border: $border-width solid getShade($quaternary-color, 60);
    border-radius: 4px;
    font-weight: 600;
    overflow: hidden;
  }
}
</style>
