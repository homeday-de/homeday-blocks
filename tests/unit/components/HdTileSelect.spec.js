import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdTileSelect from '@/components/HdTileSelect.vue';

const items = [1, 2, 3];
const value = 0;
const wrapperFactory = wrapperFactoryBuilder(HdTileSelect, {
  props: {
    value,
    items,
  },
});

describe('HdTileSelect', () => {
  it('renders component not selected', () => {
    const wrapper = wrapperFactory();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders component selected', () => {
    const wrapper = wrapperFactory({
      props: {
        value: items[0],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('accepts items entries as objects', () => {
    const objectItems = [
      {
        text: 'Yes',
        value: true,
      },
      {
        text: 'No',
        value: false,
      },
    ];
    const wrapper = wrapperFactory({
      props: {
        items: objectItems,
      },
    });
    const button = wrapper.get('button');

    expect(button.text()).toBe(objectItems[0].text);
  });

  it('emits input on item selected', () => {
    const wrapper = wrapperFactory();
    const button = wrapper.get('button');

    button.trigger('click');

    expect(wrapper.emitted().input[0][0]).toBe(items[0]);
  });

  it('applies formatter to text', () => {
    const formatter = text => (`formatted ${text}`);
    const wrapper = wrapperFactory({
      props: { formatter },
    });
    const formattedText = formatter(items[0]);
    const button = wrapper.get('button');

    expect(button.text()).toBe(formattedText);
  });

  it('allows custom new value', () => {
    const wrapper = wrapperFactory({
      props: {
        acceptNewValue: true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
