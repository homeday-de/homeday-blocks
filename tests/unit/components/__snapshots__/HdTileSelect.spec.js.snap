// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdTileSelect allows custom new value 1`] = `
<div class="field tile-select">
  <div class="tile-select__items"><button type="button" class="tile-select-item">
      1
    </button><button type="button" class="tile-select-item">
      2
    </button><button type="button" class="tile-select-item">
      3
    </button> <button type="button" class="tile-select-item tile-select-editable-item">
      +
    </button></div>
</div>
`;

exports[`HdTileSelect renders component not selected 1`] = `
<div class="field tile-select">
  <div class="tile-select__items"><button type="button" class="tile-select-item">
      1
    </button><button type="button" class="tile-select-item">
      2
    </button><button type="button" class="tile-select-item">
      3
    </button>
    <!---->
  </div>
</div>
`;

exports[`HdTileSelect renders component selected 1`] = `
<div class="field tile-select">
  <div class="tile-select__items"><button type="button" class="tile-select-item tile-select-item--selected">
      1
    </button><button type="button" class="tile-select-item">
      2
    </button><button type="button" class="tile-select-item">
      3
    </button>
    <!---->
  </div>
</div>
`;
